commands/cli.go

```diff
@@ -0,0 +1,290 @@
+package commands
+
+import (
+	"bytes"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"github.com/urfave/cli"
+	"github.com/vapor-ware/sctl/cloud"
+	"github.com/vapor-ware/sctl/utils"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"sort"
+	"strings"
+	"time"
+)
+
+// BuildContextualMenu - Assemble the CLI commands, subcommands, and flags
+// Handles the majority of the CLI interface.
+// Retuns an array of cli.Command configuration
+func BuildContextualMenu() []cli.Command {
+	return []cli.Command{
+		{
+			Name:  "add",
+			Usage: "add secret",
+			Flags: []cli.Flag{
+				cli.StringFlag{
+					Name:   "key",
+					EnvVar: "SCTL_KEY",
+					Usage:  "GCloud KMS Key URI",
+				},
+				cli.BoolFlag{
+					Name:  "no-decode",
+					Usage: "When reading the secret, do not base64 decode",
+				},
+			},
+			Action: func(c *cli.Context) error {
+
+				// Check for KMS key uri, and presence of the secrets name
+				err := validateContext(c, "add")
+				if err != nil {
+					log.Fatal(err)
+				}
+
+				var plaintext []byte
+
+				// Scan for data being piped via STDIN and favor this over alternate inputs
+				plaintext = stdinScan()
+				if plaintext == nil {
+					// NO data deteced on stdin, attempt to scan for args after keyname
+					if len(c.Args()) > 1 {
+						plaintext = []byte(c.Args()[1])
+					} else {
+						// Everything else has failed finally resort to prompting for manual input
+						plaintext = utils.UserInput()
+						// if we have nothing at this phase, log an error and abort
+						if len(plaintext) == 0 {
+							log.Fatal("Empty input detected. Aborting")
+						}
+					}
+				}
+
+				secretName := c.Args().First()
+
+				secretEncoding := ""
+				// determine if we need to base64 the raw text, defaults
+				// to true.
+				if c.Bool("no-decode") == true {
+					// skip encoding, encode as plain value
+					secretEncoding = "plain"
+				} else {
+					// encode value as base64 compressed string
+					secretEncoding = "base64"
+					plaintext = []byte(base64.StdEncoding.EncodeToString(plaintext))
+				}
+
+				// Init a KMS client
+				client := cloud.NewGCPKMS(c.String("key"))
+
+				cypher, err := client.Encrypt(plaintext)
+				if err != nil {
+					log.Fatal(err)
+				}
+				// re-encode the binary data we got back.
+				encoded := base64.StdEncoding.EncodeToString(cypher)
+				toAdd := utils.Secret{
+					Name:       strings.ToUpper(secretName),
+					Cyphertext: encoded,
+					Created:    time.Now(),
+					Encoding:   secretEncoding,
+				}
+
+				utils.AddSecret(toAdd)
+
+				return nil
+			},
+		},
+		{
+			Name:  "send",
+			Usage: "Encode/Decode a secret for copy/paste",
+			Flags: []cli.Flag{
+				cli.StringFlag{
+					Name:   "key",
+					EnvVar: "SCTL_KEY",
+					Usage:  "GCloud KMS Key URI",
+				},
```

Comment:
Since this flag is used for multiple commands, we may want to define it in a global var so all commands use the same flag. This would make it easier to update/maintain in the future.

Reply:
i disagree :) global flags for non key-required interactions are poopy

Reply:
ah, I didn't mean global flag, I meant global variable, e.g. 

```go
var FlagSctlKey = cli.StringFlag{
    Name: "key",
    EnvVar: "SCTL_KEY",
    Usage: "GCloud KMS Key URI",
}

...


		{
			Name:  "send",
			Usage: "Encode/Decode a secret for copy/paste without storing in state",
			Flags: []cli.Flag{
				FlagSctlKey,
			},
		}
```


This isn't really necessary either, so I'm a-ok with this not happening.